src/coreclr/src/pal/src/loader/module.cpp

```diff
@@ -1445,6 +1445,25 @@ static bool ShouldRedirectToCurrentLibrary(LPCSTR libraryNameOrPath)
     if (libraryNameOrPath == nullptr)
         return true;
 
+#if defined(TARGET_LINUX)
+    const char* toRedirect[] = {
+            "System.Native.so",
+            "System.IO.Compression.Native.so",
+            "System.Net.Security.Native.so",
+            "System.Security.Cryptography.Native.OpenSsl.so"
+    };
+
+    int nameLength = strlen(libraryNameOrPath);
+    int count = sizeof(toRedirect) / sizeof(toRedirect[0]);
+    for (int i = 0; i < count; ++i)
+    {
+        const char* match = toRedirect[i];
+        int matchLength = strlen(match);
```

Comment:
These lengths should be compile time constants and have no need to be computed at runtime. This should be simple to handle above.

Reply:
Right. Originally this was supposed to be a bit more dynamic. Now that only Linux is supported, this can be simplified a bit.

Reply:
Just call `strcmp` ?

Reply:
I am not sure how fast we need to be with this check. 
I can as well switch on a strlen against 4 known values and strcmp, but it might be that just strcm is good enough. These are relatively short strings.

Reply:
Actually, it looks like we try to support `some/path/System.Native.so`  as well, then knowing the length is needed.
In C# this would be `EndsWith`.

Reply:
I am going to do the following (seems like close enough to the use of const lengths):

```c
    static const char* toRedirect[] = {
        "System.Native.so",
        "System.IO.Compression.Native.so",
        "System.Net.Security.Native.so",
        "System.Security.Cryptography.Native.OpenSsl.so"
    };

    static const int toRedirectLen[] = {
        strlen(toRedirect[0]),
        strlen(toRedirect[1]),
        strlen(toRedirect[2]),
        strlen(toRedirect[3]),
    };

    int nameLength = strlen(libraryNameOrPath);
    int count = sizeof(toRedirect) / sizeof(toRedirect[0]);
    for (int i = 0; i < count; ++i)
    {
        const char* match = toRedirect[i];
        int matchLength = toRedirectLen[i];
        if (nameLength >= matchLength && strcmp(libraryNameOrPath + nameLength - matchLength, match) == 0)
            return true;
    }
```

Reply:
I do not think putting literal `{16, 31, ...}`   is a good idea. 
Also `strlen` over a literal can be used as constexpr in GCC, I think, but this is a nonstandard extension and may not be so in other compilers.

Reply:
> it looks like we try to support some/path/System.Native.so

Is it a good idea to support this? If somebody specifies a path, I would expect that we will load the stuff from that path and do not do any single-file redirects.

Reply:
Another thought: The Unix PAL may be a wrong places for these redirects. Doing the redirects for PAL would not work on Windows - there is no Unix PAL on Windows. Would it be better to do these redirects in `LoadNativeLibraryBySearch` ?

Reply:
not supporting path (just using strcmp) works. I am not sure what was the intent to support path as well. Would we ever get a path for these libraries?

`LoadNativeLibraryBySearch` seems like a better place to put the shim. Especially when supporting Windows. 
I did not know that this entry point existed. I wonder if there are caveats with placing shim there. Maybe we should do that separately.

Reply:
A lot of this came from quick&dirty prototype that was put together without much thinking. The point of this work is to do it right (tm).

Reply:
Moved the library load redirection shim to `LoadNativeLibraryBySearch`